Improve URL Handler

hqiu-at-wiris opened the pull request:
- split methods into smaller function
- split the URL into components in initialization
- return None instead of empty URLs
- remove methods not needed anymore
- remove raise exception for invalid URL flag
- refactor tests

hqiu-at-wiris committed:
refactor: Improve URL Handler

- split methods into smaller function
- split the URL into components in initialization
- return None instead of empty URLs
- remove methods not needed anymore
- remove raise exception for invalid URL flag
- refactor tests

hqiu-at-wiris committed:
fix: Add responses package version constraint to avoid moto's error

hqiu-at-wiris committed:
feat: Upgrade python version to 3.9

hqiu-at-wiris committed:
fix: Set the same python patch version for devcontainer and github actions' image

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -1,205 +1,337 @@
+            extracted_url = tldextract.extract(url=url)
+            if parsed_url.scheme:
+                scheme = parsed_url.scheme
+            subdomains = cls.process_subdomain(subdomain=extracted_url.subdomain)
```
Is this correct here? It's a bit confusing to interleave it between ifs. Put it before `path = cls.process_path(...`

dmas-at-wiris replied:
Move 97 and 102 after 94

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -1,205 +1,337 @@
-        str
-            The fully qualified domain of the `url`, following the aforementioned criteria.
+        List[str]
+            A list of subdomains up to `num_subdomains`, or an empty list if `num_subdomains` is 0.
```
what if num_subdomains is greater than the actual number of subdomains?

rroig-at-wiris replied:
If `num_subdomains` is greater than the actual number of subdomains, we return all the subdomains. We have added an explanation in the method's docstring.

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -212,152 +344,41 @@ def contains_ip(cls, url: str) -> bool:
-            return ""
+        exploded_domains = []
+        # Add 1 to include all subdomains
+        for num_subdomains in range(len(self.subdomains) + 1):
+            partial_domain = self.get_partial_domain(num_subdomains=num_subdomains)
+            if partial_domain:
+                exploded_domains.append(partial_domain)
```
Is this correct?

rroig-at-wiris replied:
It is correct, but not the most efficient since `get_partial_domain` is being called multiple times. We have refactored it to only call it once and then do successive splits on "." to obtain the different `partial_subdomain`.

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -1,205 +1,337 @@
+        return scheme, subdomains, domain, suffix, path
+
+    @classmethod
+    def process_subdomain(cls, subdomain: str) -> List[str]:
```
We might be using an incorrect definition of domain and subdomain. Since domain and subdomains included a TLD.

Let's review [standard terminology](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier). We should define somewhere what we understand right now as a subdomain based on the host and domain definition. Some [useful place](https://en.wikipedia.org/wiki/Subdomain) to gather info. 

Let's discuss about this in person.

dmas-at-wiris replied:
https://stackoverflow.com/questions/9071279/number-in-the-top-level-domain

dmas-at-wiris commented on santoku/utils/test/test_url_handler.py:
```diff
@@ -1,182 +1,308 @@
         # funcition would consider 'example' as subdomain, 'com' as domain, and 'uk' as suffix
-        ("https://example.com.uk", 0, "com.uk"),
-        ("www.example.com", 0, "example.com"),
+        ("https://example.com.uk", ("https", ["example"], "com", "uk", None)),
```
what's the meaning of this case?

rroig-at-wiris replied:
This test case wants to state that for `https://example.com.uk`, despite the `com.uk` is not a valid TLD, the URL is considered valid by `urllib.parse` and hence it is expected to be parsed into its components. We just want to make sure that even if a URL has an invalid TLD, it is parsed as seen here.

dmas-at-wiris commented on santoku/utils/test/test_url_handler.py:
```diff
@@ -1,182 +1,308 @@
+        ("https://example.com.uk", 1, ["example"]),
+        ("https://example.com", 1, []),
+        ("https://user@example.co.uk:8000", 1, []),
+        ("125.0.0.0", 2, []),
```
for context: IPs encoding format is called dotted quad notation

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -1,205 +1,337 @@
-        super().__init__(message)
+class URLHandler:
+    """
+    This class perfoms some parsing and transformation to a URL and provide methods to obtain
+    different forms of this URL.
```
```suggestion
    This class provides URL parsing and transformation functionality to different forms.
```

dmas-at-wiris commented on santoku/utils/url_handler.py:
```diff
@@ -1,205 +1,337 @@
+    This class perfoms some parsing and transformation to a URL and provide methods to obtain
+    different forms of this URL.
 
+    A URL is split into different components: `scheme`, `subdomains`, `base_domain`, `suffix` and
+    `path`.
```
Put a link to some official source (wiki, mozilla, etc) where this is described in detail

rroig-at-wiris committed:
feat: Add new methods to explode path and join paths with domains

Add tests for new methods

Added test case for TLD containing digits

Replace concat in the methods' names with join

Return empty list instead of None in explode_domain method

Improve URLHandler docstring
